File: C++/Graphs/kargers.cpp
// # Karger's Algorithm (Min Cut)

// Karger's algorithm is a randomized algorithm used to find the minimum cut in a graph. The minimum cut of a graph is defined as the smallest number of edges that, if removed, would divide the graph into two disconnected components. Karger's algorithm is based on the concept of contraction, where two randomly chosen vertices are merged into a single vertex, reducing the size of the graph. The algorithm repeats this contraction step until only two vertices (representing the two sides of the cut) remain.

// ### Here is a high-level overview of Karger's algorithm:

// 1. Start with an undirected graph.
// 2. Repeat steps 3-4 until there are only two vertices left.
// 3. Randomly select an edge from the graph.
// 4. Merge the two vertices connected by the selected edge into a single vertex, combining their edges.
// 5. Return the cut represented by the remaining two vertices.

// The algorithm's main advantage is its simplicity and efficiency. It has a high probability of finding the correct minimum cut, but it may occasionally give incorrect results due to its randomized nature. To improve the accuracy, the algorithm is usually run multiple times to increase the chances of finding the correct minimum cut.

// Karger's algorithm has a time complexity of approximately O(n^2), where n is the number of vertices in the graph. Although it has a relatively high time complexity, it is efficient in practice for graphs with a large number of vertices.

// Note: Karger's algorithm is often used as a subroutine in other algorithms and has applications in network reliability, image segmentation, and clustering problems.












#include <iostream>
#include <vector>
#include <ctime>
#include <random>

using namespace std;

// Structure to represent an edge in the graph
struct Edge {
    int src, dest;
};

// Function to find the root of a vertex
int findRoot(vector<int>& parent, int vertex) {
    if (parent[vertex] == vertex)
        return vertex;
    return findRoot(parent, parent[vertex]);
}

// Function to perform union of two subsets
void unionSets(vector<int>& parent, int x, int y) {
    int rootX = findRoot(parent, x);
    int rootY = findRoot(parent, y);
    parent[rootX] = rootY;
}

// Function to find the minimum cut using Karger's algorithm
int findMinCut(vector<Edge>& edges, int vertices) {
    vector<int> parent(vertices);
    for (int i = 0; i < vertices; i++)
        parent[i] = i;

    int remainingVertices = vertices;

    // Create a random number generator
    random_device rd;
    mt19937 rng(rd());

    // Keep contracting vertices until only 2 vertices are left
    while (remainingVertices > 2) {
        uniform_int_distribution<int> dist(0, edges.size() - 1);
        int randomEdge = dist(rng);

        int rootSrc = findRoot(parent, edges[randomEdge].src);
        int rootDest = findRoot(parent, edges[randomEdge].dest);

        if (rootSrc == rootDest)
            continue;

        unionSets(parent, rootSrc, rootDest);
        remainingVertices--;
    }

    // Count the edges crossing the cut
    int cutEdges = 0;
    for (const auto& edge : edges) {
        int rootSrc = findRoot(parent, edge.src);
        int rootDest = findRoot(parent, edge.dest);
        if (rootSrc != rootDest)
            cutEdges++;
    }

    return cutEdges;
}

int main() {
    int vertices = 4;
    int edgesCount = 5;

    vector<Edge> edges = { {0, 1}, {0, 2}, {0, 3}, {1, 3}, {2, 3} };

    int minCut = findMinCut(edges, vertices);

    cout << "Minimum cut found: " << minCut << endl;

    return 0;
}

File: Python/Graphs/kargers.py
import random

# A class to represent a subset for the Union-Find algorithm
class Subset:
    def __init__(self, parent, rank):
        self.parent = parent
        self.rank = rank

# A class to represent an unweighted edge in the graph
class Edge:
    def __init__(self, src, dest):
        self.src = src
        self.dest = dest

# A class to represent a connected, undirected, and unweighted graph
class Graph:
    def __init__(self, vertices, edges):
        self.vertices = vertices
        self.edges = edges

# A utility function to find the subset to which a particular vertex belongs (uses path compression technique)
def find(subsets, i):
    if subsets[i].parent != i:
        subsets[i].parent = find(subsets, subsets[i].parent)
    return subsets[i].parent

# A function that performs union of two subsets (uses union by rank)
def union(subsets, x, y):
    x_root = find(subsets, x)
    y_root = find(subsets, y)

    if subsets[x_root].rank < subsets[y_root].rank:
        subsets[x_root].parent = y_root
    elif subsets[x_root].rank > subsets[y_root].rank:
        subsets[y_root].parent = x_root
    else:
        subsets[y_root].parent = x_root
        subsets[x_root].rank += 1

# Karger's algorithm to find the minimum cut in a graph
def karger_min_cut(graph):
    vertices = graph.vertices
    edges = graph.edges

    # Create subsets for each vertex
    subsets = []
    for v in range(vertices):
        subsets.append(Subset(v, 0))

    # Initially, there are 'vertices' in the contracted graph
    num_vertices = vertices

    # Keep contracting vertices until there are only 2 vertices left
    while num_vertices > 2:
        # Pick a random edge
        random_edge = random.choice(edges)

        # Find the subsets of the source and destination vertices of the edge
        subset_src = find(subsets, random_edge.src)
        subset_dest = find(subsets, random_edge.dest)

        # If the source and destination vertices belong to the same subset, skip this edge
        if subset_src == subset_dest:
            continue

        # Contract the edge (merge the subsets of the source and destination vertices)
        union(subsets, subset_src, subset_dest)
        num_vertices -= 1

    # Count the number of remaining edges between the two final subsets
    min_cut = 0
    for edge in edges:
        subset_src = find(subsets, edge.src)
        subset_dest = find(subsets, edge.dest)
        if subset_src != subset_dest:
            min_cut += 1

    return min_cut

# Driver program to test the algorithm
def main():
    # Create a graph with 4 vertices and 5 edges
    V = 4
    E = 5
    graph = Graph(V, [Edge(0, 1), Edge(0, 2), Edge(0, 3), Edge(1, 2), Edge(2, 3)])

    # Run Karger's algorithm to find the minimum cut
    min_cut = karger_min_cut(graph)
    print("Minimum cut found by Karger's algorithm:", min_cut)

if __name__ == '__main__':
    main()
